embrace-android-instrumentation-network-status/src/main/kotlin/io/embrace/android/embracesdk/internal/instrumentation/network/NetworkStateDataSource.kt

```diff
@@ -0,0 +1,61 @@
+package io.embrace.android.embracesdk.internal.instrumentation.network
+
+import io.embrace.android.embracesdk.internal.arch.InstrumentationArgs
+import io.embrace.android.embracesdk.internal.arch.datasource.DataSourceImpl
+import io.embrace.android.embracesdk.internal.arch.datasource.SessionStateToken
+import io.embrace.android.embracesdk.internal.arch.limits.NoopLimitStrategy
+import io.embrace.android.embracesdk.internal.arch.schema.SchemaType.NetworkState
+import io.embrace.android.embracesdk.internal.arch.schema.SchemaType.NetworkState.Status
+import io.embrace.android.embracesdk.internal.capture.connectivity.NetworkConnectivityListener
+import io.embrace.android.embracesdk.internal.comms.delivery.NetworkStatus
+import io.embrace.android.embracesdk.internal.session.MemoryCleanerListener
+
+class NetworkStateDataSource(
+    args: InstrumentationArgs,
+) : NetworkConnectivityListener, MemoryCleanerListener, DataSourceImpl(
+    args = args,
+    limitStrategy = NoopLimitStrategy
+) {
+    private var currentState: Status = Status.UNKNOWN
+    private var sessionStateToken: SessionStateToken<Status>? = null
+
+    override fun onNetworkConnectivityStatusChanged(status: NetworkStatus) {
+        val timestamp = clock.now()
+        val oldState = currentState
+        val newState = status.toState()
+        currentState = newState
+        val currentStateToken = sessionStateToken
+        if (currentStateToken != null) {
+            captureTelemetry(inputValidation = { newState != oldState }) {
+                currentStateToken.update(
+                    updateDetectedTimeMs = timestamp,
+                    newValue = newState
+                )
+            }
+        }
+    }
+
+    override fun resetDataCaptureLimits() {
+        setNewStateSpan(currentState)
+    }
+
+    override fun cleanCollections() {
```

Comment:
Not required for this PR, just stating my intention that we should combine these two functions on DataSource in future as they effectively do the same thing